### PR TITLE
#11 - crud for speakers

### PR DESCRIPTION
**Pull request by MrHDOLEK**
- add crud
- mini refactor
- bind interface for storage
- fix docker-compose and readme how to run project

### CODE REVIEW

File: src/Http/Controllers/SpeakersController.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    public function create(): View
+    {
+        return view("speakers.create");
+    }
+
+    public function edit(Speaker $speaker): View
+    {
+        return view("speakers.edit")
+            ->with("speakers", $speaker);
```
Comment (MichalMyskow) [inline src/Http/Controllers/SpeakersController.php:31]:
```suggestion
            ->with("speaker", $speaker);
```

File: src/Http/Routing/WebRouting.php [hunk 2]
```diff
@@ -30,5 +31,14 @@ public function wire(): void
             $this->router->put("/organizations/{organization}", "update")->name("organizations.update");
             $this->router->delete("/organizations/{organization}", "destroy")->name("organizations.destroy");
         });
+
+        $this->router->controller(SpeakersController::class)->group(function (): void {
+            $this->router->get("/speakers", "index")->name("speakers");
+            $this->router->post("/speakers", "store")->name("speakers.store");
+            $this->router->get("/speakers/create", "create")->name("speakers.create");
+            $this->router->get("/speakers/{meetup}/edit", "edit")->name("speakers.edit");
```
Comment (MichalMyskow) [inline src/Http/Routing/WebRouting.php:39]:
```suggestion
            $this->router->get("/speakers/{speaker}/edit", "edit")->name("speakers.edit");
```

File: resources/views/speakers/edit.blade.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                        <x-input-error for="name"/>
+                    </div>
+                    <div>
+                        <label for="description">Description:</label>
+                        <textarea name="description" id="description" cols="30" rows="4">{{ old('description', $speaker->description) }}</textarea>
+                        <x-input-error for="description"/>
+                    </div>
+                    <div>
+                        <label for="avatar">Avatar:</label>
+                        <input type="text" id="avatar" name="avatar" value="{{ old('avatar', $speaker->avatar) }}">
```
Comment (MichalMyskow) [inline resources/views/speakers/edit.blade.php:23]:
```suggestion
                        <input type="file" id="avatar" name="avatar">
```

**Review by MichalMyskow — CHANGES_REQUESTED** [review]

File: resources/views/speakers/create.blade.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
+@extends('layouts.app')
+
+@section('content')
+    <div>
+        <div>
+            <h1>Add Speakers</h1>
+            @auth
```
Comment (kamilpiech97) [inline resources/views/speakers/create.blade.php:7]: Whats doing here this `@auth`?

File: resources/views/speakers/edit.blade.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                        <x-input-error for="avatar"/>
+                    </div>
+                    <div>
+                        <label for="linkedin_url">Linkedin url:</label>
+                        <input type="text" id="linkedin_url" name="linkedin_url" value="{{ old('linkedin_url', $speaker->linkedin_url) }}">
```
Comment (kamilpiech97) [inline resources/views/speakers/edit.blade.php:28]:
```suggestion
                        <input type="text" id="linkedin_url" name="linkedin_url" value="{{ old('linkedin_url', $speaker->linkedinUrl) }}">
```
Use camel case trait from previous PR.

File: resources/views/speakers/index.blade.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+            @auth
+                <a href="{{ route('speakers.create') }}">Add new speakers</a>
+            @endauth
+            @if ($speakers->count())
+                @foreach ($speakers as $speaker)
+                    <div>
+                        {{ $speaker->name }}
+                        {{ $speaker->description }}
+                        {{ $speaker->avatar }}
+                        {{ $speaker->linkedin_url }}
```
Comment (kamilpiech97) [inline resources/views/speakers/index.blade.php:16]:
```suggestion
                        {{ $speaker->linkedinUrl }}
```
The same as above.

File: src/Http/Controllers/MeetupController.php [hunk 7]
```diff
@@ -14,7 +14,7 @@ class MeetupController extends Controller
 {
     public function index(): View
     {
-        $meetups = Meetup::latest()->with(["user"])->paginate(20);
+        $meetups = Meetup::query()->latest()->with(["user"])->paginate(20);
```
Comment (kamilpiech97) [inline src/Http/Controllers/MeetupController.php:17]: Move maybe number of items per page into some variable placed in config file.

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: src/Http/Controllers/MeetupController.php (hunk 7 above)
Comment (MrHDOLEK) [reply src/Http/Controllers/MeetupController.php:17]: Not better a move this option to front .And make it possible for the user to choose pagination 20, 40, 60, e.g.

File: resources/views/speakers/create.blade.php (hunk 4 above)
Comment (MrHDOLEK) [reply resources/views/speakers/create.blade.php:7]: so far @auth is doing that a user who is not logged in will not enter it

Comment (MrHDOLEK) [reply resources/views/speakers/create.blade.php:7]: Just like the other blade

Comment (kamilpiech97) [reply resources/views/speakers/create.blade.php:7]: maybe middleware will be better ? :>

File: src/Http/Controllers/MeetupController.php (hunk 7 above)
Comment (kamilpiech97) [reply src/Http/Controllers/MeetupController.php:17]: Maybe but starting value should be loaded.

File: resources/views/speakers/edit.blade.php [hunk 8]
```diff
@@ -20,17 +20,17 @@
...
                     <div>
                         <label for="avatar">Avatar:</label>
-                        <input type="text" id="avatar" name="avatar" value="{{ old('avatar', $speaker->avatar) }}">
+                        <input type="file" id="avatar" name="avatar">
                         <x-input-error for="avatar"/>
                     </div>
                     <div>
                         <label for="linkedin_url">Linkedin url:</label>
-                        <input type="text" id="linkedin_url" name="linkedin_url" value="{{ old('linkedin_url', $speaker->linkedin_url) }}">
+                        <input type="text" id="linkedin_url" name="linkedin_url" value="{{ old('linkedinUrl', $speaker->linkedinUrl) }}">
```
[the remark this answers is not in the document]
Comment (MichalMyskow) [reply resources/views/speakers/edit.blade.php:28]:
```suggestion
                        <input type="text" id="linkedin_url" name="linkedin_url" value="{{ old('linkedin_url', $speaker->linkedinUrl) }}">
```
Function old() is using the name attribute from input.

File: resources/views/speakers/create.blade.php (hunk 4 above)
Comment (MrHDOLEK) [reply resources/views/speakers/create.blade.php:7]: better but @Majkulnn  still working for auth and not have middleware with working

File: resources/views/speakers/edit.blade.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (kamilpiech97) [reply resources/views/speakers/edit.blade.php:28]: yes, sorry my mistake.

File: src/Models/Speaker.php [hunk 9]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Blumilk\Meetup\Core\Models;
+
+use Blumilk\Meetup\Core\Formats;
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
```
Comment (MichalMyskow) [inline src/Models/Speaker.php:9]:
```suggestion
```
`CamelCaseAttributes` trait was not working. Speaker model was extending default Eloquent model.

File: src/Http/Controllers/SpeakersController.php [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        $service->storeFile("speakers", $request->file("avatar"));
+        Speaker::query()->create($request->validated());
+
+        return redirect()->route("speakers");
+    }
+
+    public function update(UpdateSpeakerRequest $request, Speaker $speaker): RedirectResponse
+    {
+        $speaker->update($request->validated());
```
Comment (MichalMyskow) [inline src/Http/Controllers/SpeakersController.php:44]: We need to think about storing and updating files.

**Review by MichalMyskow — CHANGES_REQUESTED** [review]

File: resources/views/speakers/edit.blade.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (kamilpiech97) [reply resources/views/speakers/edit.blade.php]: Sorry for the fail with `old()`.

File: src/Http/Controllers/SpeakersController.php (hunk 10 above)
Comment (MrHDOLEK) [reply src/Http/Controllers/SpeakersController.php:44]: Maybe discuse for this topic for nex weekly @MichalMyskow  ?

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: src/Providers/StoreFileProvider.php [hunk 11]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Blumilk\Meetup\Core\Providers;
+
+use Blumilk\Meetup\Core\Contracts\StoreFile;
+use Blumilk\Meetup\Core\Services\StoreFileService as Service;
```
[the remark this answers is not in the document]
Comment (MrHDOLEK) [reply src/Providers/StoreFileProvider.php:8]: Because the code is short, concise and you don't need to break it down

File: src/Providers/StoreFileProvider.php [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+use Illuminate\Support\ServiceProvider;
+
+class StoreFileProvider extends ServiceProvider
```
[the remark this answers is not in the document]
Comment (MrHDOLEK) [reply src/Providers/StoreFileProvider.php:11]: I think that it is needed, because we have file support in several places and there will be more of them. I try binding a concreate instance for a default laravel behavior and not working

File: src/Providers/StoreFileProvider.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (krzysztofrewak) [reply src/Providers/StoreFileProvider.php]: I still don't get it. In case of any other service bindng, your approach will be at least ambiguous. I would stay with imported real class name `StoreFileService` and use it in binder. 

Btw. you can use `$this->app->bind()` instead of helper `app()`.

File: readme.md [hunk 13]
```diff
@@ -2,16 +2,20 @@
 
 ### Development
 ### Run env for Mac/Linux
+
 1. `$ make install`
 2. `$ make start`
-3. if the artisan starts serving the website, then you can do the migrations `$ make db-create`
+3. `touch database/database.sqlite`
+4. `$ make bash` and run this command `composer install`
+6. `$ make db-create`
```
Comment (mtracz) [inline readme.md:10]: You can create one `make` command for all points.

File: readme.md [hunk 14]
```diff
@@ -2,16 +2,20 @@
...
 ### Run env for Windows
 
-1. `docker-compose up`
+1. `docker-compose up -d`
 2. `cp .env.example .env`
-3. if the artisan starts serving the website, then you can do the migrations 
-   `docker-compose exec php php artisan migrate`
+3. `touch database/database.sqlite`
+4. `docker-compose exec php ash` and run this command `composer install`
+5. `docker-compose exec php php artisan migrate`
```
Comment (mtracz) [inline readme.md:18]: On Windows, there is also `make` package.
http://gnuwin32.sourceforge.net/packages/make.htm

File: src/Http/Requests/Speaker/Rules/GithubLinkRules.php [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+declare(strict_types=1);
+
+namespace Blumilk\Meetup\Core\Http\Requests\Speaker\Rules;
+
+use Blumilk\Meetup\Core\Http\Requests\BaseRules;
+
+class GithubLinkRules extends BaseRules
+{
+    protected static array $rules = [
+        "regex:/http(s)?:\/\/(www\.)?github\.com\/.+/i",
```
Comment (mtracz) [inline src/Http/Requests/Speaker/Rules/GithubLinkRules.php:12]: Maybe `starts_with` will be better than this regex
https://laravel.com/docs/9.x/validation#rule-starts-with

**Review by mtracz — CHANGES_REQUESTED** [review]

File: resources/views/speakers/index.blade.php [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+                @foreach ($speakers as $speaker)
+                    <div>
+                        {{ $speaker->name }}
+                        {{ $speaker->description }}
+                        {{ $speaker->avatar }}
+                        {{ $speaker->linkedinUrl }}
+                        {{ $speaker->githubUrl }}
+
+                        <a href="{{ route('speakers.edit', $speaker) }}">Edit</a>
+                        <form action="{{ route('speakers.destroy', $speaker) }}" method="post">
```
Comment (kamilpiech97) [inline resources/views/speakers/index.blade.php:20]: Any confirmation for that action? Smth like [that](https://developer.mozilla.org/en-US/docs/Web/API/Window/confirm).

File: src/Http/Controllers/SpeakersController.php [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+
+        return redirect()->route("speakers");
```
Comment (kamilpiech97) [inline src/Http/Controllers/SpeakersController.php:46]: Can we update the avatar?

File: src/Http/Requests/Speaker/UpdateSpeakerRequest.php [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+use Blumilk\Meetup\Core\Http\Requests\Speaker\Rules\LinkedinLinkRules;
+use Illuminate\Foundation\Http\FormRequest;
+
+class UpdateSpeakerRequest extends FormRequest
+{
+    public function rules(): array
+    {
+        return [
+            "name" => "required",
+            "description" => "nullable",
```
Comment (kamilpiech97) [inline src/Http/Requests/Speaker/UpdateSpeakerRequest.php:18]:
```suggestion
            "description" => ["nullable", "max:255"],
```
Maybe upper limit?

File: resources/views/meetups/index.blade.php [hunk 19]
```diff
@@ -20,7 +20,7 @@
                         <form action="{{ route('meetups.destroy', $meetup) }}" method="post">
                             @csrf
                             @method('DELETE')
-                            <button type="submit">Delete</button>
+                            <button type="submit"  onclick="return confirm('Sure Want Delete?')">Delete</button>
```
Comment (kamilpiech97) [inline resources/views/meetups/index.blade.php:23]:
```suggestion
                            <button type="submit"  onclick="return confirm('Delete meetup? This operation is irreversible.')">Delete</button>
```

**Review by kamilpiech97 — DISMISSED** [review]

File: .gitignore [hunk 20]
```diff
@@ -12,11 +12,12 @@
 .env.backup
 .phpunit.result.cache
 docker-compose.override.yml
-composer.lock
```
Comment (krzysztofrewak) [inline .gitignore:15]: Why did you remove lock from gitignore?

File: src/Http/Middleware/RedirectIfAuthenticated.php [hunk 21]
```diff
@@ -15,7 +15,8 @@ class RedirectIfAuthenticated
     public function __construct(
         protected Redirector $redirector,
         protected AuthManager $authManager,
-    ) {}
+    ) {
+    }
```
Comment (krzysztofrewak) [inline src/Http/Middleware/RedirectIfAuthenticated.php:19]: You are apparently not using 0.10 version of our codestyle package.

File: src/Http/Requests/Speaker/UpdateSpeakerRequest.php [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+use Blumilk\Meetup\Core\Http\Requests\Speaker\Rules\GithubLinkRules;
+use Blumilk\Meetup\Core\Http\Requests\Speaker\Rules\LinkedinLinkRules;
+use Illuminate\Foundation\Http\FormRequest;
+
+class UpdateSpeakerRequest extends FormRequest
+{
+    public function rules(): array
+    {
+        return [
+            "name" => "required",
```
Comment (krzysztofrewak) [inline src/Http/Requests/Speaker/UpdateSpeakerRequest.php:17]: Maybe you could use arrays of strings everywhere?

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: src/Http/Requests/Speaker/Rules/GithubLinkRules.php [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+declare(strict_types=1);
+
+namespace Blumilk\Meetup\Core\Http\Requests\Speaker\Rules;
+
+use Blumilk\Meetup\Core\Http\Requests\BaseRules;
+
+class GithubLinkRules extends BaseRules
+{
+    protected static array $rules = [
+        "starts_with:https://github.com/",
```
Comment (mtracz) [inline src/Http/Requests/Speaker/Rules/GithubLinkRules.php:12]: To handle more valid URLs you should provide more prefixes:
```
"starts_with:github.com/,https://github.com/"
```

File: src/Http/Requests/Speaker/Rules/LinkedinLinkRules.php [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+declare(strict_types=1);
+
+namespace Blumilk\Meetup\Core\Http\Requests\Speaker\Rules;
+
+use Blumilk\Meetup\Core\Http\Requests\BaseRules;
+
+class LinkedinLinkRules extends BaseRules
+{
+    protected static array $rules = [
+        "starts_with:https://www.linkedin.com/",
```
Comment (mtracz) [inline src/Http/Requests/Speaker/Rules/LinkedinLinkRules.php:12]: Let's define more prefixes:
```
"starts_with:linkedin.com/,www.linkedin.com/,https://linkedin.com/,https://www.linkedin.com/",
```

**Review by mtracz — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 20 above)
Comment (MrHDOLEK) [reply .gitignore:15]: because I don't see the sense of keeping lock in ignor with development .

File: src/Http/Requests/Speaker/UpdateSpeakerRequest.php (hunk 22 above)
Comment (MrHDOLEK) [reply src/Http/Requests/Speaker/UpdateSpeakerRequest.php:17]: what do you mean ?

File: Makefile [hunk 25]
```diff
@@ -1,31 +1,54 @@
...
+	docker-compose exec php php artisan key:generate
+
+init: ## Setup for project
+	make install
+	make sqlite-create
+	make start
+	make key-generate
+	make composer-install
+	make sqlite-create
+	make db-create
```
[the remark this answers is not in the document]
Comment (MrHDOLEK) [reply Makefile]: Yes but a create comand help .Which works by taking the comment from the first line and displaying all commands when listing them

File: src/Http/Requests/Speaker/UpdateSpeakerRequest.php (hunk 22 above)
Comment (kamilpiech97) [inline src/Http/Requests/Speaker/UpdateSpeakerRequest.php:17]:
```suggestion
            "name" => ["required"],
```

**Review by mtracz — APPROVED** [review]

**Review by kamilpiech97 — APPROVED** [review]

File: .gitignore (hunk 20 above)
Comment (krzysztofrewak) [reply .gitignore:15]: It's a library, not a project. It's useless.

**Review by krzysztofrewak — APPROVED** [review]

**Review by kamilpiech97 — APPROVED** [review]

**Review by kamilpiech97 — APPROVED** [review]

**Review by MichalMyskow — APPROVED** [review]
